File: ImlacPDS1/SourcesObjC/NetSession.m
// NetSession.m — UDP LAN multiplayer for Imlac PDS-1 iOS
// Ported from NetSession.swift (originally NetSession.java)
#import "NetSession.h"
#import <sys/socket.h>
#import <netinet/in.h>
#import <arpa/inet.h>
#import <unistd.h>

#define NET_PKT 32
#define NET_PEER_TIMEOUT 6.0

typedef NS_ENUM(NSInteger, NetRole)   { NetRoleNone, NetRoleHost, NetRoleGuest };
typedef NS_ENUM(NSInteger, NetStatus) { NetStatusIdle, NetStatusHosting, NetStatusSearching, NetStatusConnected, NetStatusDisconnected };

@interface NetSession () {
    NetRole _role;
    NetStatus _status;
    NSString *_peerAddrStr;

    int _gameSocket;
    int _bcastSocket;
    BOOL _running;
    dispatch_queue_t _netQueue;
    NSMutableArray<NSData *> *_sendQueue;
    NSLock *_sendLock;
    NSDate *_lastPeerTime;
}
@end

@implementation NetSession

+ (NSInteger)PORT_GAME  { return 7474; }
+ (NSInteger)PORT_BCAST { return 7475; }

- (instancetype)init {
    self = [super init];
    if (self) {
        _role = NetRoleNone;
        _status = NetStatusIdle;
        _gameSocket = -1;
        _bcastSocket = -1;
        _running = NO;
        _netQueue = dispatch_queue_create("net.session", DISPATCH_QUEUE_SERIAL);
        dispatch_set_target_queue(_netQueue, dispatch_get_global_queue(QOS_CLASS_USER_INTERACTIVE, 0));
        _sendQueue = [NSMutableArray array];
        _sendLock = [[NSLock alloc] init];
        _peerMazeX = 1; _peerMazeY = 1; _peerMazeDir = 0; _peerMazeHp = 3; _peerMazeScore = 0;
    }
    return self;
}

- (NSString *)peerAddr { return _peerAddrStr; }

#pragma mark - Host

- (void)host:(NSInteger)seed {
    [self stop];
    _mazeSeed = seed; _role = NetRoleHost; _myId = 0; _status = NetStatusHosting; _running = YES;
    __weak NetSession *weakSelf = self;
    dispatch_async(_netQueue, ^{ [weakSelf hostLoop]; });
}

- (void)hostLoop {
    _gameSocket  = [self makeUDPSocketPort:(int)[NetSession PORT_GAME]  broadcast:NO];
    _bcastSocket = [self makeUDPSocketPort:(int)[NetSession PORT_BCAST] broadcast:YES];
    if (!(_gameSocket >= 0 && _bcastSocket >= 0)) { [self fail]; return; }

    struct timeval timeout = { .tv_sec = 0, .tv_usec = 300000 };
    setsockopt(_bcastSocket, SOL_SOCKET, SO_RCVTIMEO, &timeout, sizeof(timeout));
    setsockopt(_gameSocket,  SOL_SOCKET, SO_RCVTIMEO, &timeout, sizeof(timeout));

    uint8_t buf[NET_PKT];
    while (_running && _status == NetStatusHosting) {
        struct sockaddr_storage srcAddr;
        socklen_t srcLen = sizeof(srcAddr);
        memset(&srcAddr, 0, sizeof(srcAddr));
        ssize_t n = recvfrom(_bcastSocket, buf, NET_PKT, 0, (struct sockaddr *)&srcAddr, &srcLen);
        if (n == NET_PKT && buf[0] == 'D') {
            _peerAddrStr = [self addrToString:&srcAddr];
            NSData *welcome = [self makePacketType:'W' idv:0 demo:0 kbd:0 seed:(int)_mazeSeed];
            [self sendTo:_gameSocket data:welcome addr:&srcAddr addrLen:srcLen];
        }
        ssize_t n2 = recvfrom(_gameSocket, buf, NET_PKT, 0, (struct sockaddr *)&srcAddr, &srcLen);
        if (n2 == NET_PKT && buf[0] == 'J') {
            _peerAddrStr = [self addrToString:&srcAddr];
            _status = NetStatusConnected;
            _lastPeerTime = [NSDate date];
            NSInteger seedCopy = _mazeSeed;
            __weak NetSession *weakSelf = self;
            dispatch_async(dispatch_get_main_queue(), ^{
                [weakSelf.eventListener onConnectedAsHost:YES seed:seedCopy];
            });
        }
    }
    if (_status == NetStatusConnected) { [self gameLoop]; }
    else { [self fail]; }
}

#pragma mark - Guest

- (void)discover {
    [self stop];
    _role = NetRoleGuest; _myId = 1; _status = NetStatusSearching; _running = YES;
    __weak NetSession *weakSelf = self;
    dispatch_async(_netQueue, ^{ [weakSelf guestLoop]; });
}

- (void)guestLoop {
    _gameSocket = [self makeUDPSocketPort:(int)[NetSession PORT_GAME] broadcast:YES];
    if (!(_gameSocket >= 0)) { [self fail]; return; }

    struct timeval timeout = { .tv_sec = 0, .tv_usec = 300000 };
    setsockopt(_gameSocket, SOL_SOCKET, SO_RCVTIMEO, &timeout, sizeof(timeout));

    NSData *bcastData = [self makePacketType:'D' idv:1 demo:0 kbd:0 seed:0];
    NSDate *deadline = [NSDate dateWithTimeIntervalSinceNow:30];
    uint8_t buf[NET_PKT];

    while (_running && _status == NetStatusSearching && [[NSDate date] compare:deadline] == NSOrderedAscending) {
        [self broadcastSend:_gameSocket data:bcastData port:(int)[NetSession PORT_BCAST]];

        struct sockaddr_storage srcAddr;
        socklen_t srcLen = sizeof(srcAddr);
        memset(&srcAddr, 0, sizeof(srcAddr));
        ssize_t n = recvfrom(_gameSocket, buf, NET_PKT, 0, (struct sockaddr *)&srcAddr, &srcLen);
        if (n == NET_PKT && buf[0] == 'W') {
            _peerAddrStr = [self addrToString:&srcAddr];
            _mazeSeed = [self decodeSeed:buf offset:12];
            NSData *join = [self makePacketType:'J' idv:1 demo:0 kbd:0 seed:0];
            [self sendTo:_gameSocket data:join addr:&srcAddr addrLen:srcLen];
            _status = NetStatusConnected;
            _lastPeerTime = [NSDate date];
            NSInteger seedCopy = _mazeSeed;
            __weak NetSession *weakSelf = self;
            dispatch_async(dispatch_get_main_queue(), ^{
                [weakSelf.eventListener onConnectedAsHost:NO seed:seedCopy];
            });
        }
    }
    if (_status == NetStatusConnected) { [self gameLoop]; }
    else { [self fail]; }
}

#pragma mark - Game loop

- (void)gameLoop {
    struct timeval timeout = { .tv_sec = 0, .tv_usec = 200000 };
    setsockopt(_gameSocket, SOL_SOCKET, SO_RCVTIMEO, &timeout, sizeof(timeout));
    uint8_t buf[NET_PKT];

    while (_running) {
        [_sendLock lock];
        NSArray<NSData *> *pending = [_sendQueue copy];
        [_sendQueue removeAllObjects];
        [_sendLock unlock];
        for (NSData *pkt in pending) { [self sendToPeer:_gameSocket data:pkt]; }

        struct sockaddr_storage srcAddr;
        socklen_t srcLen = sizeof(srcAddr);
        memset(&srcAddr, 0, sizeof(srcAddr));
        ssize_t n = recvfrom(_gameSocket, buf, NET_PKT, 0, (struct sockaddr *)&srcAddr, &srcLen);
        if (n == NET_PKT) {
            _lastPeerTime = [NSDate date];
            [self handlePacket:buf];
        }

        if (_lastPeerTime && [[NSDate date] timeIntervalSinceDate:_lastPeerTime] > NET_PEER_TIMEOUT) {
            [self fail];
            return;
        }
    }
}

- (void)handlePacket:(uint8_t *)p {
    char type = (char)p[0];
    __weak NetSession *weakSelf = self;
    switch (type) {
        case 'S': {
            NSInteger demo = p[2], kbd = (p[3] << 8) | p[4];
            dispatch_async(dispatch_get_main_queue(), ^{
                [weakSelf.eventListener onPeerSyncDemoIdx:demo keyboard:kbd];
            });
            break;
        }
        case 'M': {
            NSInteger x = p[5], y = p[6], dir = p[7], hp = p[8], sc = (p[9] << 8) | p[10];
            _peerMazeX = x; _peerMazeY = y; _peerMazeDir = dir; _peerMazeHp = hp; _peerMazeScore = sc;
            dispatch_async(dispatch_get_main_queue(), ^{
                [weakSelf.eventListener onPeerMazeStateX:x y:y dir:dir hp:hp score:sc];
            });
            break;
        }
        case 'B': {
            NSInteger dir = p[11];
            dispatch_async(dispatch_get_main_queue(), ^{
                [weakSelf.eventListener onPeerBulletDir:dir];
            });
            break;
        }
        case 'K': {
            dispatch_async(dispatch_get_main_queue(), ^{
                [weakSelf.eventListener onPeerKilled];
            });
            break;
        }
        case 'C': {
            NSString *msg = [self decodeChat:p offset:16];
            dispatch_async(dispatch_get_main_queue(), ^{
                [weakSelf.chatListener onChatMessageFrom:@"OPP" msg:msg];
            });
            break;
        }
        case 'X':
            [self fail];
            break;
        default: break;
    }
}

#pragma mark - Send (thread-safe)

- (void)sendSync:(NSInteger)demoIdx keyboard:(NSInteger)keyboard {
    if (![self isConnected]) return;
    uint8_t p[NET_PKT]; memset(p, 0, sizeof(p));
    p[0] = 'S'; p[1] = (uint8_t)_myId; p[2] = (uint8_t)demoIdx;
    p[3] = (uint8_t)((keyboard >> 8) & 0xFF); p[4] = (uint8_t)(keyboard & 0xFF);
    [self enqueue:[NSData dataWithBytes:p length:NET_PKT]];
}

- (void)sendMazeStateX:(NSInteger)x y:(NSInteger)y dir:(NSInteger)dir hp:(NSInteger)hp score:(NSInteger)score {
    if (![self isConnected]) return;
    uint8_t p[NET_PKT]; memset(p, 0, sizeof(p));
    p[0] = 'M'; p[1] = (uint8_t)_myId;
    p[5] = (uint8_t)x; p[6] = (uint8_t)y; p[7] = (uint8_t)dir; p[8] = (uint8_t)hp;
    p[9] = (uint8_t)((score >> 8) & 0xFF); p[10] = (uint8_t)(score & 0xFF);
    [self enqueue:[NSData dataWithBytes:p length:NET_PKT]];
}

- (void)sendBullet:(NSInteger)dir {
    if (![self isConnected]) return;
    uint8_t p[NET_PKT]; memset(p, 0, sizeof(p));
    p[0] = 'B'; p[1] = (uint8_t)_myId; p[11] = (uint8_t)dir;
    [self enqueue:[NSData dataWithBytes:p length:NET_PKT]];
}

- (void)sendKill {
    if (![self isConnected]) return;
    uint8_t p[NET_PKT]; memset(p, 0, sizeof(p));
    p[0] = 'K'; p[1] = (uint8_t)_myId;
    [self enqueue:[NSData dataWithBytes:p length:NET_PKT]];
}

- (void)sendChat:(NSString *)text {
    if (![self isConnected]) return;
    uint8_t p[NET_PKT]; memset(p, 0, sizeof(p));
    p[0] = 'C'; p[1] = (uint8_t)_myId;
    const char *utf8 = [text UTF8String];
    NSInteger len = MIN((NSInteger)strlen(utf8), 15);
    for (NSInteger i = 0; i < len; i++) { p[16 + i] = (uint8_t)utf8[i]; }
    [self enqueue:[NSData dataWithBytes:p length:NET_PKT]];
    __weak NetSession *weakSelf = self;
    dispatch_async(dispatch_get_main_queue(), ^{
        [weakSelf.chatListener onChatMessageFrom:@"ME" msg:text];
    });
}

- (void)enqueue:(NSData *)d {
    [_sendLock lock];
    [_sendQueue addObject:d];
    [_sendLock unlock];
}

#pragma mark - Socket utils

- (int)makeUDPSocketPort:(int)port broadcast:(BOOL)broadcast {
    int fd = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
    if (fd < 0) return -1;
    int on = 1;
    setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, &on, sizeof(on));
    if (broadcast) { setsockopt(fd, SOL_SOCKET, SO_BROADCAST, &on, sizeof(on)); }
    struct sockaddr_in addr;
    memset(&addr, 0, sizeof(addr));
    addr.sin_family = AF_INET;
    addr.sin_port = htons((uint16_t)port);
    addr.sin_addr.s_addr = INADDR_ANY;
    bind(fd, (struct sockaddr *)&addr, sizeof(addr));
    return fd;
}

- (void)sendTo:(int)fd data:(NSData *)data addr:(struct sockaddr_storage *)addr addrLen:(socklen_t)addrLen {
    sendto(fd, data.bytes, data.length, 0, (struct sockaddr *)addr, addrLen);
}

- (void)sendToPeer:(int)fd data:(NSData *)data {
    if (!_peerAddrStr) return;
    struct sockaddr_in addr;
    memset(&addr, 0, sizeof(addr));
    addr.sin_family = AF_INET;
    addr.sin_port = htons((uint16_t)[NetSession PORT_GAME]);
    addr.sin_addr.s_addr = inet_addr([_peerAddrStr UTF8String]);
    sendto(fd, data.bytes, data.length, 0, (struct sockaddr *)&addr, sizeof(addr));
}

- (void)broadcastSend:(int)fd data:(NSData *)data port:(int)port {
    struct sockaddr_in addr;
    memset(&addr, 0, sizeof(addr));
    addr.sin_family = AF_INET;
    addr.sin_port = htons((uint16_t)port);
    addr.sin_addr.s_addr = INADDR_BROADCAST;
    sendto(fd, data.bytes, data.length, 0, (struct sockaddr *)&addr, sizeof(addr));
}

- (NSString *)addrToString:(struct sockaddr_storage *)addr {
    char buf[INET_ADDRSTRLEN];
    struct sockaddr_in *sin = (struct sockaddr_in *)addr;
    inet_ntop(AF_INET, &(sin->sin_addr), buf, INET_ADDRSTRLEN);
    return [NSString stringWithUTF8String:buf];
}

- (NSData *)makePacketType:(char)type idv:(int)idv demo:(int)demo kbd:(int)kbd seed:(int)seed {
    uint8_t p[NET_PKT]; memset(p, 0, sizeof(p));
    p[0] = (uint8_t)type; p[1] = (uint8_t)idv; p[2] = (uint8_t)demo;
    p[3] = (uint8_t)((kbd >> 8) & 0xFF); p[4] = (uint8_t)(kbd & 0xFF);
    p[12] = (uint8_t)((seed >> 24) & 0xFF); p[13] = (uint8_t)((seed >> 16) & 0xFF);
    p[14] = (uint8_t)((seed >> 8) & 0xFF);  p[15] = (uint8_t)(seed & 0xFF);
    return [NSData dataWithBytes:p length:NET_PKT];
}

- (NSInteger)decodeSeed:(uint8_t *)p offset:(int)offset {
    return ((NSInteger)p[offset] << 24) | ((NSInteger)p[offset+1] << 16) | ((NSInteger)p[offset+2] << 8) | (NSInteger)p[offset+3];
}

- (NSString *)decodeChat:(uint8_t *)p offset:(int)offset {
    char buf[16];
    int i = 0;
    for (; i < 15; i++) {
        uint8_t b = p[offset + i];
        if (b == 0) break;
        buf[i] = (char)b;
    }
    buf[i] = 0;
    NSString *s = [NSString stringWithUTF8String:buf];
    return s ?: @"";
}

- (void)fail {
    _status = NetStatusDisconnected;
    __weak NetSession *weakSelf = self;
    dispatch_async(dispatch_get_main_queue(), ^{
        [weakSelf.eventListener onDisconnected];
    });
    [self closeAll];
}

- (void)stop {
    _running = NO;
    [self closeAll];
    _status = NetStatusIdle;
    _role = NetRoleNone;
}

- (void)closeAll {
    if (_gameSocket >= 0) { close(_gameSocket); _gameSocket = -1; }
    if (_bcastSocket >= 0) { close(_bcastSocket); _bcastSocket = -1; }
}

- (BOOL)isConnected { return _status == NetStatusConnected; }

@end

File: ImlacPDS1/Sources/NetSession.swift
// NetSession.swift — UDP LAN multiplayer for Imlac PDS-1 iOS
// Ported from NetSession.java

import Foundation
import Network

// ── Event listener protocol ───────────────────────────────
protocol NetSessionEventListener: AnyObject {
    func onConnected(asHost: Bool, seed: Int)
    func onPeerMazeState(_ x:Int,_ y:Int,_ dir:Int,_ hp:Int,_ score:Int)
    func onPeerSync(_ demoIdx:Int,_ keyboard:Int)
    func onPeerBullet(_ dir:Int)
    func onPeerKilled()
    func onDisconnected()
}

protocol NetSessionChatListener: AnyObject {
    func onChatMessage(from: String, msg: String)
}

// ── NetSession ────────────────────────────────────────────
final class NetSession {

    static let PORT_GAME  = 7474
    static let PORT_BCAST = 7475
    private static let PKT = 32
    private static let PEER_TIMEOUT: TimeInterval = 6.0

    enum Role   { case none, host, guest }
    enum Status { case idle, hosting, searching, connected, disconnected }

    private(set) var role:   Role   = .none
    private(set) var status: Status = .idle
    private(set) var peerAddr: String?

    // Peer state
    var peerMazeX=1,peerMazeY=1,peerMazeDir=0,peerMazeHp=3,peerMazeScore=0
    var mazeSeed: Int = 0
    var myId: Int = 0

    weak var eventListener: NetSessionEventListener?
    weak var chatListener:  NetSessionChatListener?

    // Networking (using BSD sockets via GCD)
    private var gameSocket:  Int32 = -1
    private var bcastSocket: Int32 = -1
    private var running = false
    private var netQueue = DispatchQueue(label: "net.session", qos: .userInteractive)
    private var sendQueue = [Data]()
    private var sendLock  = NSLock()
    private var lastPeerTime: Date?

    // ─────────────────────────────────────────────────────────
    //  HOST
    // ─────────────────────────────────────────────────────────
    func host(_ seed: Int) {
        stop(); mazeSeed=seed; role = .host; myId=0; status = .hosting; running=true
        netQueue.async { self.hostLoop() }
    }

    private func hostLoop() {
        gameSocket  = makeUDPSocket(port: NetSession.PORT_GAME,  broadcast: false)
        bcastSocket = makeUDPSocket(port: NetSession.PORT_BCAST, broadcast: true)
        guard gameSocket >= 0 && bcastSocket >= 0 else { fail(); return }

        // Discovery phase — listen for guest broadcast
        let timeout = timeval(tv_sec: 0, tv_usec: 300_000)
        setsockopt(bcastSocket, SOL_SOCKET, SO_RCVTIMEO, &timeout, socklen_t(MemoryLayout<timeval>.size))
        setsockopt(gameSocket,  SOL_SOCKET, SO_RCVTIMEO, &timeout, socklen_t(MemoryLayout<timeval>.size))

        var buf = [UInt8](repeating:0, count:NetSession.PKT)
        while running && status == .hosting {
            // Check broadcast socket for 'D' (discover)
            var srcAddr = sockaddr_storage()
            var srcLen  = socklen_t(MemoryLayout<sockaddr_storage>.size)
            let n = withUnsafeMutablePointer(to: &srcAddr) {
                $0.withMemoryRebound(to: sockaddr.self, capacity:1) {
                    recvfrom(bcastSocket, &buf, NetSession.PKT, 0, $0, &srcLen)
                }
            }
            if n == NetSession.PKT && buf[0] == UInt8(ascii:"D") {
                peerAddr = addrToString(&srcAddr)
                let welcome = makePacket(type:"W", id:0, demo:0, kbd:0, seed:mazeSeed)
                sendTo(gameSocket, data: welcome, addr: &srcAddr, addrLen: srcLen)
            }
            // Check game socket for 'J' (join ack)
            let n2 = withUnsafeMutablePointer(to: &srcAddr) {
                $0.withMemoryRebound(to: sockaddr.self, capacity:1) {
                    recvfrom(gameSocket, &buf, NetSession.PKT, 0, $0, &srcLen)
                }
            }
            if n2 == NetSession.PKT && buf[0] == UInt8(ascii:"J") {
                peerAddr = addrToString(&srcAddr)
                status = .connected
                lastPeerTime = Date()
                DispatchQueue.main.async { self.eventListener?.onConnected(asHost:true, seed:self.mazeSeed) }
            }
        }
        if status == .connected { gameLoop() }
        else { fail() }
    }

    // ─────────────────────────────────────────────────────────
    //  GUEST
    // ─────────────────────────────────────────────────────────
    func discover() {
        stop(); role = .guest; myId=1; status = .searching; running=true
        netQueue.async { self.guestLoop() }
    }

    private func guestLoop() {
        gameSocket = makeUDPSocket(port: NetSession.PORT_GAME, broadcast: true)
        guard gameSocket >= 0 else { fail(); return }

        let timeout = timeval(tv_sec: 0, tv_usec: 300_000)
        setsockopt(gameSocket, SOL_SOCKET, SO_RCVTIMEO, &timeout, socklen_t(MemoryLayout<timeval>.size))

        let bcastData = makePacket(type:"D", id:1, demo:0, kbd:0, seed:0)
        let deadline  = Date().addingTimeInterval(30)
        var buf = [UInt8](repeating:0, count:NetSession.PKT)

        while running && status == .searching && Date() < deadline {
            broadcastSend(gameSocket, data: bcastData, port: NetSession.PORT_BCAST)

            var srcAddr = sockaddr_storage()
            var srcLen  = socklen_t(MemoryLayout<sockaddr_storage>.size)
            let n = withUnsafeMutablePointer(to: &srcAddr) {
                $0.withMemoryRebound(to: sockaddr.self, capacity:1) {
                    recvfrom(gameSocket, &buf, NetSession.PKT, 0, $0, &srcLen)
                }
            }
            if n == NetSession.PKT && buf[0] == UInt8(ascii:"W") {
                peerAddr = addrToString(&srcAddr)
                mazeSeed = decodeSeed(buf, offset: 12)
                let join = makePacket(type:"J", id:1, demo:0, kbd:0, seed:0)
                sendTo(gameSocket, data: join, addr: &srcAddr, addrLen: srcLen)
                status = .connected
                lastPeerTime = Date()
                DispatchQueue.main.async { self.eventListener?.onConnected(asHost:false, seed:self.mazeSeed) }
            }
        }
        if status == .connected { gameLoop() }
        else { fail() }
    }

    // ─────────────────────────────────────────────────────────
    //  GAME LOOP
    // ─────────────────────────────────────────────────────────
    private func gameLoop() {
        let timeout = timeval(tv_sec: 0, tv_usec: 200_000)
        setsockopt(gameSocket, SOL_SOCKET, SO_RCVTIMEO, &timeout, socklen_t(MemoryLayout<timeval>.size))
        var buf = [UInt8](repeating:0, count:NetSession.PKT)

        while running {
            // Drain send queue
            sendLock.lock(); let pending = sendQueue; sendQueue.removeAll(); sendLock.unlock()
            for pkt in pending { sendToPeer(gameSocket, data: pkt) }

            // Receive
            var srcAddr = sockaddr_storage(); var srcLen = socklen_t(MemoryLayout<sockaddr_storage>.size)
            let n = withUnsafeMutablePointer(to: &srcAddr) {
                $0.withMemoryRebound(to: sockaddr.self, capacity:1) {
                    recvfrom(gameSocket, &buf, NetSession.PKT, 0, $0, &srcLen)
                }
            }
            if n == NetSession.PKT { lastPeerTime = Date(); handlePacket(buf) }

            // Peer timeout
            if let lp = lastPeerTime, Date().timeIntervalSince(lp) > NetSession.PEER_TIMEOUT {
                fail(); return
            }
        }
    }

    private func handlePacket(_ p: [UInt8]) {
        let type = Character(UnicodeScalar(p[0]))
        switch type {
        case "S":
            let demo = Int(p[2]), kbd = (Int(p[3])<<8) | Int(p[4])
            DispatchQueue.main.async { self.eventListener?.onPeerSync(demo, kbd) }
        case "M":
            let x=Int(p[5]),y=Int(p[6]),dir=Int(p[7]),hp=Int(p[8]),sc=(Int(p[9])<<8)|Int(p[10])
            peerMazeX=x;peerMazeY=y;peerMazeDir=dir;peerMazeHp=hp;peerMazeScore=sc
            DispatchQueue.main.async { self.eventListener?.onPeerMazeState(x,y,dir,hp,sc) }
        case "B":
            let dir=Int(p[11])
            DispatchQueue.main.async { self.eventListener?.onPeerBullet(dir) }
        case "K":
            DispatchQueue.main.async { self.eventListener?.onPeerKilled() }
        case "C":
            let msg = decodeChat(p, offset: 16)
            DispatchQueue.main.async { self.chatListener?.onChatMessage(from:"OPP", msg:msg) }
        case "X":
            fail()
        default: break
        }
    }

    // ─────────────────────────────────────────────────────────
    //  SEND (thread-safe)
    // ─────────────────────────────────────────────────────────
    func sendSync(_ demoIdx: Int, _ keyboard: Int) {
        guard isConnected() else { return }
        var p = [UInt8](repeating:0, count:NetSession.PKT)
        p[0]=UInt8(ascii:"S"); p[1]=UInt8(myId); p[2]=UInt8(demoIdx)
        p[3]=UInt8((keyboard>>8)&0xFF); p[4]=UInt8(keyboard&0xFF)
        enqueue(Data(p))
    }
    func sendMazeState(_ x:Int,_ y:Int,_ dir:Int,_ hp:Int,_ score:Int) {
        guard isConnected() else { return }
        var p = [UInt8](repeating:0, count:NetSession.PKT)
        p[0]=UInt8(ascii:"M"); p[1]=UInt8(myId)
        p[5]=UInt8(x);p[6]=UInt8(y);p[7]=UInt8(dir);p[8]=UInt8(hp)
        p[9]=UInt8((score>>8)&0xFF);p[10]=UInt8(score&0xFF)
        enqueue(Data(p))
    }
    func sendBullet(_ dir: Int) {
        guard isConnected() else { return }
        var p = [UInt8](repeating:0, count:NetSession.PKT)
        p[0]=UInt8(ascii:"B"); p[1]=UInt8(myId); p[11]=UInt8(dir)
        enqueue(Data(p))
    }
    func sendKill() {
        guard isConnected() else { return }
        var p = [UInt8](repeating:0, count:NetSession.PKT)
        p[0]=UInt8(ascii:"K"); p[1]=UInt8(myId)
        enqueue(Data(p))
    }
    func sendChat(_ text: String) {
        guard isConnected() else { return }
        var p = [UInt8](repeating:0, count:NetSession.PKT)
        p[0]=UInt8(ascii:"C"); p[1]=UInt8(myId)
        let bytes = Array(text.utf8.prefix(15))
        for (i,b) in bytes.enumerated() { p[16+i]=b }
        enqueue(Data(p))
        DispatchQueue.main.async { self.chatListener?.onChatMessage(from:"ME", msg:text) }
    }
    private func enqueue(_ d: Data) { sendLock.lock(); sendQueue.append(d); sendLock.unlock() }

    // ─────────────────────────────────────────────────────────
    //  SOCKET UTILS
    // ─────────────────────────────────────────────────────────
    private func makeUDPSocket(port: Int, broadcast: Bool) -> Int32 {
        let fd = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP)
        guard fd >= 0 else { return -1 }
        var on: Int32 = 1
        setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, &on, socklen_t(MemoryLayout<Int32>.size))
        if broadcast { setsockopt(fd, SOL_SOCKET, SO_BROADCAST, &on, socklen_t(MemoryLayout<Int32>.size)) }
        var addr = sockaddr_in()
        addr.sin_family = sa_family_t(AF_INET)
        addr.sin_port   = in_port_t(UInt16(port).bigEndian)
        addr.sin_addr   = in_addr(s_addr: INADDR_ANY)
        withUnsafePointer(to: &addr) {
            $0.withMemoryRebound(to: sockaddr.self, capacity:1) {
                Darwin.bind(fd, $0, socklen_t(MemoryLayout<sockaddr_in>.size))
            }
        }
        return fd
    }

    private func sendTo(_ fd: Int32, data: Data, addr: inout sockaddr_storage, addrLen: socklen_t) {
        var d = data
        withUnsafeMutablePointer(to: &addr) {
            $0.withMemoryRebound(to: sockaddr.self, capacity:1) { sa in
                d.withUnsafeMutableBytes { ptr in
                    sendto(fd, ptr.baseAddress, data.count, 0, sa, addrLen)
                }
            }
        }
    }

    private func sendToPeer(_ fd: Int32, data: Data) {
        guard let peer = peerAddr else { return }
        var addr = sockaddr_in()
        addr.sin_family = sa_family_t(AF_INET)
        addr.sin_port   = in_port_t(UInt16(NetSession.PORT_GAME).bigEndian)
        addr.sin_addr.s_addr = inet_addr(peer)
        withUnsafePointer(to: &addr) {
            $0.withMemoryRebound(to: sockaddr.self, capacity:1) { sa in
                var d = data
                d.withUnsafeMutableBytes { ptr in
                    sendto(fd, ptr.baseAddress, data.count, 0, sa, socklen_t(MemoryLayout<sockaddr_in>.size))
                }
            }
        }
    }

    private func broadcastSend(_ fd: Int32, data: Data, port: Int) {
        var addr = sockaddr_in()
        addr.sin_family = sa_family_t(AF_INET)
        addr.sin_port   = in_port_t(UInt16(port).bigEndian)
        addr.sin_addr.s_addr = INADDR_BROADCAST
        withUnsafePointer(to: &addr) {
            $0.withMemoryRebound(to: sockaddr.self, capacity:1) { sa in
                var d = data
                d.withUnsafeMutableBytes { ptr in
                    sendto(fd, ptr.baseAddress, data.count, 0, sa, socklen_t(MemoryLayout<sockaddr_in>.size))
                }
            }
        }
    }

    private func addrToString(_ addr: inout sockaddr_storage) -> String {
        var buf = [CChar](repeating:0, count:Int(INET_ADDRSTRLEN))
        withUnsafePointer(to: &addr) {
            $0.withMemoryRebound(to: sockaddr_in.self, capacity:1) {
                var a = $0.pointee.sin_addr
                inet_ntop(AF_INET, &a, &buf, socklen_t(INET_ADDRSTRLEN))
            }
        }
        return String(cString: buf)
    }

    private func makePacket(type: String, id: Int, demo: Int, kbd: Int, seed: Int) -> Data {
        var p = [UInt8](repeating:0, count:NetSession.PKT)
        p[0]=type.utf8.first ?? 0; p[1]=UInt8(id); p[2]=UInt8(demo)
        p[3]=UInt8((kbd>>8)&0xFF); p[4]=UInt8(kbd&0xFF)
        p[12]=UInt8((seed>>24)&0xFF); p[13]=UInt8((seed>>16)&0xFF)
        p[14]=UInt8((seed>>8)&0xFF);  p[15]=UInt8(seed&0xFF)
        return Data(p)
    }

    private func decodeSeed(_ p: [UInt8], offset: Int) -> Int {
        (Int(p[offset])<<24)|(Int(p[offset+1])<<16)|(Int(p[offset+2])<<8)|Int(p[offset+3])
    }

    private func decodeChat(_ p: [UInt8], offset: Int) -> String {
        let bytes = p[offset...].prefix(while: { $0 != 0 })
        return String(bytes: bytes, encoding: .utf8) ?? ""
    }

    private func fail() {
        status = .disconnected
        DispatchQueue.main.async { self.eventListener?.onDisconnected() }
        closeAll()
    }

    func stop() {
        running = false; closeAll(); status = .idle; role = .none
    }

    private func closeAll() {
        if gameSocket  >= 0 { Darwin.close(gameSocket);  gameSocket  = -1 }
        if bcastSocket >= 0 { Darwin.close(bcastSocket); bcastSocket = -1 }
    }

    func isConnected() -> Bool { status == .connected }
}
